### PR TITLE
chore: add release workflow

### PR DESCRIPTION
**Pull request by techouse**
Addresses #51

I have made the `main.yml` workflow reusable and used it as a test job upon which the publish job depends.

1. a `pull_request` of a branch named `release/*` will trigger a dry run
2. a `push` will trigger a deployment if the tags match this pattern:
  - `'envied-v[0-9]+.[0-9]+.[0-9]+*'`
  - `'envied_generator-v[0-9]+.[0-9]+.[0-9]+*'`

I didn't use melos here as I saw no need and because melos uses `dart pub publish` [under the hood anyway](https://melos.invertase.dev/guides/automated-releases#publishing).

I have left a comment for you in the workflow where you will have to create an environment with the name `pub.dev` (or you can name it differently) just as it's outlined here in [this guide](https://dart.dev/tools/pub/automated-publishing#configuring-automated-publishing-from-github-actions-on-pubdev).

### CODE REVIEW

**Comment by petercinibulk** [comment]
@techouse what would be the workflow for making a release with this action? It seems like you would need to have a tag or a release branch. You also would need to manually update the version and changelog before this is run.

I am thinking that either you need to make a "release" PR or we make it part of the contributing PR step to have contributors run the Melos script to generate a new version and changelog before the PR is merged.

I am curious what your thoughts are?

**Comment by techouse** [comment]
> what would be the workflow for making a release with this action? 

You push a tag, just like one of these that you already have here https://github.com/petercinibulk/envied/tags.

If you make a `release/` branch PR it will run a `dart pub publish --dry-run`, essentially testing if it can publish without publishing.

**Comment by petercinibulk** [comment]
Okay. Those tags and changelogs are created with the '''melos version''' command btw. I'll get that env added and merge this to test

**Comment by techouse** [comment]
Hold off on merging it just yet. I realised, that providing the environment for the dry run job is not desired and one can not just conditionally add it. I'll refactor this into a reusable workflow and then call it from 2 separate workflow files with their own environments respectively.

**Comment by techouse** [comment]
To avoid any future confusion I've renamed `main.yml` to `test.yml` (as that is essentially what it does).

I have also changed the `publish.yml` workflow to only trigger on pushing tags:
```yaml
on:
  push:
    tags:
      - 'envied-v[0-9]+.[0-9]+.[0-9]+*'
      - 'envied_generator-v[0-9]+.[0-9]+.[0-9]+*'
```
for example:

- `envied-v1.2.3`
- `envied-v4.5.6+780`
- `envied_generator-v1.2.3`
- `envied_generator-v4.5.6+780`

Separately I have created a `publish_dry_run.yml` which as the name implies only does a [dry run](https://dart.dev/tools/pub/cmd/pub-lish#--dry-run-or--n) and is triggered on any pull request that has the branch name:
```yaml
on:
  pull_request:
    branches:
      - 'release/v**'
      - 'release/envied/v**'
      - 'release/envied_generator/v**'
```
for example:

- `release/v1.2.3`
- `release/envied/v4.5.6`
- `release/envied_generator/v7.8.0+123`

**Review by petercinibulk — APPROVED** [review]

**Pull request merged**
